### PR TITLE
bloc pattern on stream

### PR DESCRIPTION
**Pull request by sergei-kabak-idf**
Added bloc architecture on stream

### CODE REVIEW

**Review by yuliya-klescheva-idf — APPROVED** [review]

**Review by alexei-kruk-idf — APPROVED** [review]

File: packages/domain/lib/use_case/palindrome_use_case.dart [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import 'package:domain/use_case/use_case.dart';
 
-class Palindrome implements UseCase {
+class PalindromeUseCaseImpl implements UseCase {
   bool _checkPalindrome(String text, int first, int last) {
     if (text.isEmpty | text.contains(RegExp(r'[0-9,^\s*]'))) {
       return false;
```
Comment (sergei-mikhailovskii-idf) [inline packages/domain/lib/use_case/palindrome_use_case.dart]: in future, call palindrome usecase via call method

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  final StreamController<PalindromeData> _controllerStream =
+      StreamController<PalindromeData>();
+  final StreamController<HomeBlocEvent> _eventController =
+      StreamController<HomeBlocEvent>();
+
+  TextEditingController get controllerText => _controllerText;
+
+  Sink<PalindromeData> get _inputData => _controllerStream.sink;
+
+  Sink<HomeBlocEvent> get inputCheckPalindrome => _eventController.sink;
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/bloc/palindrome_bloc.dart:25]: why do u need sinks here if u can add events directly to the StreamController?

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+import 'package:domain/use_case/palindrome_use_case.dart';
+import 'package:flutter/material.dart';
+import 'package:presentation/screen/bloc/palindrome_data.dart';
+import 'package:presentation/screen/bloc/palindrome_event.dart';
+
+abstract class Bloc {
+  void dispose();
+}
+
+class PalindromeBloc extends Bloc {
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/bloc/palindrome_bloc.dart:12]: try to separate abstraction and implementation how I showed u (https://github.dev/sergei-mikhailovskii-idf/flutter_lab/blob/master/packages/presentation/lib/screen/home_bloc.dart). in this case controllerText, inputCheckPalindrome and some other fields will be in abstraction

File: packages/presentation/lib/screen/bloc/palindrome_event.dart [hunk 4]
```diff
@@ -0,0 +1 @@
+enum HomeBlocEvent { checkPalindrome}
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/bloc/palindrome_event.dart]: name enums in the uppercase

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+  PalindromeBloc() {
+    _outputCheckPalindrome.listen(_listenEvent);
+  }
+
+  void _listenEvent(HomeBlocEvent event) {
+    switch (event) {
+      case HomeBlocEvent.checkPalindrome:
+        data.isPalindrome = palindrome.call(
+            _controllerText.text, 0, _controllerText.text.length);
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/bloc/palindrome_bloc.dart]: formatting? or it is in a formatted way?

File: packages/presentation/lib/screen/home_page.dart [hunk 6]
```diff
@@ -29,25 +48,14 @@ class _HomePageState extends State<HomePage> {
             ),
             textAlign: TextAlign.center,
           ),
-          Text(answer.toString()),
+          Text(snapshot.data?.isPalindrome.toString() ?? 'null'),
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/home_page.dart:51]: if u will call Text(snapshot.data?.isPalindrome.toString()) (without handling null case), won't null be printed?

File: pubspec.lock [hunk 7]
```diff
@@ -109,6 +109,13 @@ packages:
       url: "https://pub.dartlang.org"
```
Comment (sergei-mikhailovskii-idf) [inline pubspec.lock:109]: gitignore

**Review by sergei-mikhailovskii-idf — APPROVED** [review]

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  Sink<PalindromeData> get _inputData => _controllerStream.sink;
+
+  Sink<HomeBlocEvent> get inputCheckPalindrome => _eventController.sink;
+
+  Stream<PalindromeData> get outputData => _controllerStream.stream;
+
+  Stream<HomeBlocEvent> get _outputCheckPalindrome => _eventController.stream;
+
+  PalindromeBloc() {
+    _outputCheckPalindrome.listen(_listenEvent);
```
Comment (sergei-mikhailovskii-idf) [inline packages/presentation/lib/screen/bloc/palindrome_bloc.dart:32]: it's ok, that widget passes events to bloc, but it'd be better to make the communications between them via calling bloc methods

File: packages/presentation/lib/screen/bloc/palindrome_event.dart (hunk 4 above)
Comment (sergei-kabak-idf) [reply packages/presentation/lib/screen/bloc/palindrome_event.dart]: fixed

File: pubspec.lock (hunk 7 above)
Comment (sergei-kabak-idf) [reply pubspec.lock:109]: fixed

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart (hunk 8 above)
Comment (sergei-kabak-idf) [reply packages/presentation/lib/screen/bloc/palindrome_bloc.dart:32]: I see. I'm correcting that code.

File: packages/domain/lib/use_case/palindrome_use_case.dart (hunk 1 above)
Comment (sergei-kabak-idf) [reply packages/domain/lib/use_case/palindrome_use_case.dart]: fixed

File: packages/presentation/lib/screen/bloc/palindrome_bloc.dart (hunk 5 above)
Comment (sergei-kabak-idf) [reply packages/presentation/lib/screen/bloc/palindrome_bloc.dart]: fixed

**Pull request merged**